README.md
# Summary

Habitat4J intends to simplify configuration management in a multiple
server, multiple application environment.  However, it is designed to be
easy to use -- even in a single application on a single server.

Habitat4J provides property management based on a two component model.  
The first component, called the "Server Identity," provides a system-wide
context for which applications load their properties.  The second
component, called the "Property List," provides an application-wide
property infrastructure that will load entries based on information in the
Server Identity component.

# Feature Summary
- XML-based configuration
- Property loading from filesystem or classpath resource
- Ability to set System and Security properties
- Name/value pairs, arrays, hashes, Javabean, Javabean array, and Javabean hash Property Values
- Optional property reloading at runtime - explicitly, on-the-fly, or per time interval
- Optional XSD Validation
- Optional Encoded and/or Encrypted Property Values
- Optional support for log4j

# Uses for Habitat4J include:

- Any environment running multiple applications</li>
- Staging, unit testing, and production server environments</li>
- Server farm environments</li>

# Introduction

Habitat4J is an XML-based property file management system for Java.
It relies on a two component model:

- ServerIdentity Component a.k.a. XML-based "ServerIdentity File"

  - This file defines a server-wide set of parameters that all
  applications residing on that server will utilize to load
  a subset of properties in the PropertyList (see below).
  
- PropertyList Component(s) a.k.a. XML-based "PropertyList File(s)"

  - This file contains Property values that are organized by context.
  Contexts will be loaded depending on the ServerIdentity.

# Detailed Overview

Habitat4J provides the following for your application:

- The following Property types are supported:

  - Standard Name/Value Pair String Properties
  - String Array Properties (list of Strings associated with a property
    name)
  - String Hash Properties (hash of Strings associated with a property
    name, with entries that can be looked up by an additional key)
  - Javabean Properties (reuse your own or create new)
  - Javabean Array Properties (list of Javabeans associated with a
    property name)
  - Javabean Hash Properties (hash of Javabeans associated with a
    proeprty name, with entries that can be looked up by an additional
    key)
  - System Properties
  - Security Properties

- The option is available to decode and/or decrypt values (in any of the
  above Property types).  Decoder and Decryptor classes are not
  provided by Habitat4J, but an interface is given to write your own
  implementations that utilize popular or proprietary encoding and
  encryption mechanisms.  One sample Decoder and one sample Decryptor
  class is provided with the Habitat4J JUnit tests.

- The option is available to have all ServerIdentity and PropertyList files
  validated against XSD schemas.

- The option to safely reload property files based on changing timestamps,
  either by explicitly telling Habitat4J to check for a new timestamp,
  or automatically on-the-fly, or after an interval of time has passed.
  
# Code Overview

- Straight-forward SAX-based parsing implementation.

- Support for System.out/System.err and Log4J-based logging.

- An abstract class implementation so your application can customize
  Habitat4J.
  
- The Habitat4JBootstrapServlet, available in the Habitat4JWeb
  distribution, which provides an implementation for initializing
  or "boostrapping" Habitat4J for a J2EE web application.
  
# Dependencies

- The habitat4j implementations depend on the following 3rd-party JARs:

  - xerces XML API (commonly called xml-apis.jar)
  xerces Implementation (commonly called xercesImpl.jar)
  SAX Implementation (typically included with xerces)
  log4j (optional)

- In addition, the habitat4j-x.y.z-junit.jar also depends on the following
3rd-party JAR:

  - junit
  
- The habitat4jweb-x.y.z.jar included with the Habitat4J distributions
depends on an implementation of javax.servlet.*, which is typically
included in a Servlet or J2EE container implementation.


CHANGELOG.md
# 2019-09-10 [1.0.0] jpy
- Moved to Github productivityorg/habitat4j with Maven build setup.

# 200?-??-?? [1.0.0rc19] jpy
- No idea when this version was released.  It's been a while. :)

# 2007-03-09 [1.0.0rc18] jpy
- Corrected a bug in PropertyList.isPropertyABean() to avoid a Null-
  PointerException when globalPropertyBeanDefinitions are defined but
  there are no propertyBeanDefinitions.  Method now checks both
  Hashtables.
  
# 2007-02-25 [1.0.0rc17] jpy
- Viva Habitat4J Release Candidate 17!  :)
- Added add/remove/clearReloadablePropertyList(s) methods to
  AbstractPropertyListManager and PropertyListManager to control what
  Property Lists are allowed to be reloaded.  If none are specified, all
  Property Lists are allowed to be reloaded.  (The global features for
  reloading must be turned on for any of these to work.)
- Added "fileSystemPropertyListReloadPropertyLists" property-array to
  ApplicationPropertyListLoader (used by Habitat4JBootstrap), which
  uses the addReloadablePropertyList method in PropertyListManager to
  specify which Property Lists are allowed to be reloaded.
- Added ApplicationG JUnit test cases to test Habitat4JBootstrap and
  the above changes.
- After a reload occurs, reload handlers not defined in the 
  <definitions> section that are available in the "old"
  PropertyList will be transferred to the "new" PropertyList.  This can
  be disabled with the "featurePropertyListReloadDisableHandlerTransfer"
  feature.
- Added getReloadEventHandler() and hasReloadEventHandler() methods to
  PropertyList.
  
# 2006-08-16 [1.0.0rc16] jpy
- Viva Habitat4J Release Candidate 16!  :)
- Added "instanceOS" fixed parameter to ServerIdentity, which contains
  one of three values:  "windows", "unix", or "other".  The information
  is obtained via the InstanceOsHelper, which uses the System
  properties "os.name" value to detect the operating system under which
  the JVM is running.
  
# 2006-07-08 [1.0.0rc15] jpy
- Viva Habitat4J Release Candidate 15!  :)
- Modified Habitat4JBootstrapServletContextListener to work with
  JDK 1.3- by putting a toString() on a rethrown Exception.
- Fixed issue with showing encoded and/or encrypted values with
  PropertyList.toString() when using Property Bean Hashes.  Also wasn't
  passing in a null name parameter to the DecryptorIF implementations,
  and this has been fixed.
- Added support for serialization-generated clones of Property Bean
  objects as an alternative to the standard reflection-based mechanism.
  Added an attribute for <definition><property-bean ... copyMethod="x">
  where "x" = "reflect" or "serialize", which allows users of
  the convenient getPropertyBeanCopy(), getPropertyBeanArrayCopy(), and
  getPropertyBeanHashCopy() methods to select whether reflection or
  serialization should be used to produce the clone.  The "serialize" 
  option can only be used to copy Objects that are fully serializable,
  i.e. all components and superclasses must also be fully serializable.
  If this attribute is left out, reflection is used by default.
  The property-list.xsd (and alternate property-list) schemas have been
  modified with the new copyMethod attribute.
  
# 2006-03-15 [1.0.0rc14] jpy
- Viva Habitat4J Release Candidate 14!  :)
- Refactored PropertyList.toString() to use multi-line output for
  large property definitions, such as those with many array or hash
  entries.
- Refactored System and Security properties to not log an "error"
  when setting a property value matching what's already set.

# 2006-02-14 [1.0.0rc13] jpy
- Viva Habitat4J Release Candidate 13!  :)
- Added two wildcard capabilities to context parameters:
  <context parameter="*"> .. </context>
    This context will match if any value is provided for that
    context definition.
  <context parameter="!value"> .. </context>
    This context will match if any value does *not* match for
    that context definition.
- Added PROPERTY_LIST_FEATURE_SUPPRESS_PROPERTY_BEAN_COPY_WARNINGS
  feature to suppress WARN logging that occurs when PropertyBeans
  cannot be copied entirely.

# 2006-01-25 [1.0.0rc12] jpy
- Viva Habitat4J Release Candidate 12!  :)
- In AbstractBaseHandler, added secondary attempt to instantiate the
  XMLReader if XMLReaderFactory.createXMLReader(..) throws an
  Exception.  Might come in handy when JDK 1.4- or JAXB blocks
  Xerces.
- Added "pragma" to definitions section in PropertyList.  Can be used
  to "switch" quickly between multiple configs.  PropertyList pragma
  definitions are checked first, then ServerIdentity pragma
  definitions.  The PROPERTY_LIST_FEATURE_SERVER_IDENTITY_PRAGMA_OVERRIDE
  feature needs to be set to allow PropertyList pragma definitions
  to override ServerIdentity pragma definitions.

# 2005-11-15 [1.0.0rc11] jpy
- Viva Habitat4J Release Candidate 11!  :)
- Bug Fix:  The <property-array> element had issues with mixed
  contexts.  Fix involved checking for "current context" more often.
- Added logging for the name of a classPath resource when an exception
  occurs.
- Added createServerIdentityPath() static helper method to
  Habitat4JBootstrap, which creates a semicolon-delimited path for
  server-identity files that may appear in different locations on
  different systems.

# 2005-10-06 [1.0.0rc10] jpy
- Viva Habitat4J Release Candidate 10!  :)
- Bug Fix:  The <context> "prefix" attribute was handled incorrectly.
  Anyone using relatively complex <context> sections using (a) the
  "prefix" attribute and (b) contextual attributes ("environment," etc.)
  should upgrade to 1.0.0rc10 immediately.
- Improved the PropertyBean copying code to use any existing "hasField"
  methods (no parameters and a boolean or Boolean return value) to decide
  whether a field should be copied.  If the bean does not have a
  "has" method corresponding to the "get"/"set" methods, the field will be
  copied by default.  Useful if you wish to avoid copying a field within a
  PropertyBean when using the *Copy property-retrieval methods.
  

Habitat4j/src/main/org/productivity/java/habitat4j/common/Habitat4JConstants.java
package org.productivity.java.habitat4j.common;

/**
 * Contains constant (public static final) entries used by Habitat4J.
 * 
 * <p>## LICENSE INFORMATION ##</p>
 * 
 * <p>This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.</p>
 *
 * <p>This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.</p>
 *
 * <p>You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 
 * USA</p>
 * 
 * @author Justin Yunke &lt;habitat4j@productivity.org&gt;
 * @version $Id: Habitat4JConstants.java,v 1.43 2007/02/25 04:23:10 cvs Exp $
 */
public interface Habitat4JConstants {
	public static final String HABITAT4J_VERSION											= "1.0.0rc18 2007-03-09 jpy";
	
	public static final String SAX_DRIVER_SYSTEM_PROPERTY_NAME								= "org.xml.sax.driver";
	
	public static final String HABITAT4J_SAX_DRIVER_SYSTEM_PROPERTY_NAME 					= "org.productivity.java.habitat4j.sax.driver";
	public static final String HABITAT4J_SAX_DRIVER_DEFAULT									= "org.apache.xerces.parsers.SAXParser";

	public static final String HABITAT4J_WINDOWS_OS_NAME_LIST_PROPERTY_NAME					= "org.productivity.java.habitat4j.windows.os.name.list";
	public static final String HABITAT4J_UNIX_OS_NAME_LIST_PROPERTY_NAME					= "org.productivity.java.habitat4j.unix.os.name.list";

	public static final String HABITAT4J_SAX_VALIDATION_FEATURE								= "http://xml.org/sax/features/validation";
	public static final String HABITAT4J_SAX_SCHEMA_FEATURE									= "http://apache.org/xml/features/validation/schema";
	public static final String HABITAT4J_SAX_XSD_LOCATION									= "http://apache.org/xml/properties/schema/" + "external-noNamespaceSchemaLocation";

	public static final String HABITAT4J_SERVER_IDENTITY_XSD_RESOURCE_PATH_DEFAULT			= "org/productivity/java/habitat4j/common/xsd/server-identity.xsd";
	public static final String HABITAT4J_PROPERTY_LIST_XSD_RESOURCE_PATH_DEFAULT			= "org/productivity/java/habitat4j/common/xsd/property-list.xsd";
	
	public static final String HABITAT4J_GENERIC_VALUE_DEFAULT								= "default";	
	
	public static final String HABITAT4J_GENERIC_VALUE_EMPTY_STRING							= "";	

	public static final String HABITAT4J_GENERIC_VALUE_TRUE									= "true";
	public static final String HABITAT4J_GENERIC_VALUE_ON									= "on";
	public static final String HABITAT4J_GENERIC_VALUE_ONE									= "1";
	public static final String HABITAT4J_GENERIC_VALUE_YES									= "yes";
	
	public static final String HABITAT4J_GENERIC_VALUE_FALSE								= "false";
	public static final String HABITAT4J_GENERIC_VALUE_OFF									= "off";
	public static final String HABITAT4J_GENERIC_VALUE_ZERO									= "0";
	public static final String HABITAT4J_GENERIC_VALUE_NO									= "no";
	
	public static final String HABITAT4J_SERVER_IDENTITY_FILE_PATH_SYSTEM_PROPERTY_NAME		= "org.productivity.java.habitat4j.serverIdentityFilePath";
	public static final String HABITAT4J_SERVER_IDENTITY_FILE_PATH_DEFAULT					= "/server-identity.xml";
	
	public static final String HABITAT4J_APPNAME_DEFAULT									= "default";
	public static final String HABITAT4J_APPNAME_ATTRIBUTE_NAME								= "name";

	public static final String HABITAT4J_ROLE_ATTRIBUTE_NAME								= "name";

	public static final String HABITAT4J_PRAGMA_ATTRIBUTE_NAME								= "name";	
	public static final String HABITAT4J_PRAGMA_ATTRIBUTE_VALUE								= "value";
	
	public static final String HABITAT4J_PROPERTY_VALUE_SUBSTITUTION_PREFIX					= "${";
	public static final String HABITAT4J_PROPERTY_VALUE_SUBSTITUTION_DELIMITER				= ":";
	public static final String HABITAT4J_PROPERTY_VALUE_SUBSTITUTION_SUFFIX					= "}";

	public static final String HABITAT4J_DEFINITIONS_PACKAGE								= "package";
	
	public static final String HABITAT4J_PROPERTY_LIST_NAME_DEFAULT							= "default";	
	public static final String HABITAT4J_PROPERTY_NAME_ATTRIBUTE_NAME						= "name";	
	public static final String HABITAT4J_PROPERTY_VALUE_ATTRIBUTE_NAME						= "value";
	public static final String HABITAT4J_PROPERTY_DECODING_ATTRIBUTE_NAME					= "decoderId";
	public static final String HABITAT4J_PROPERTY_DECRYPTING_ATTRIBUTE_NAME					= "decryptorId";

	public static final String HABITAT4J_PROPERTY_ARRAY_INSTANCE_ATTRIBUTE_NAME				= "instance";

	public static final String HABITAT4J_PROPERTY_ARRAY_INSTANCE_ATTRIBUTE_VALUE_REPLACE	= "replace";
	public static final String HABITAT4J_PROPERTY_ARRAY_INSTANCE_ATTRIBUTE_VALUE_PREPEND	= "prepend";
	public static final String HABITAT4J_PROPERTY_ARRAY_INSTANCE_ATTRIBUTE_VALUE_APPEND		= "append";
	public static final String HABITAT4J_PROPERTY_ARRAY_INSTANCE_ATTRIBUTE_VALUE_DEFAULT	= HABITAT4J_PROPERTY_ARRAY_INSTANCE_ATTRIBUTE_VALUE_REPLACE; 

	public static final String HABITAT4J_PROPERTY_HASH_INSTANCE_ATTRIBUTE_NAME				= "instance";
	public static final String HABITAT4J_PROPERTY_HASH_TYPE_ATTRIBUTE_NAME					= "type";
	public static final String HABITAT4J_PROPERTY_HASH_KEY_ATTRIBUTE_NAME					= "key";

	public static final String HABITAT4J_PROPERTY_HASH_INSTANCE_ATTRIBUTE_VALUE_REPLACE		= "replace";
	public static final String HABITAT4J_PROPERTY_HASH_INSTANCE_ATTRIBUTE_VALUE_REVISE		= "revise";
	public static final String HABITAT4J_PROPERTY_HASH_INSTANCE_ATTRIBUTE_VALUE_DEFAULT		= HABITAT4J_PROPERTY_HASH_INSTANCE_ATTRIBUTE_VALUE_REPLACE; 

	public static final String HABITAT4J_PROPERTY_BEAN_DESCRIPTOR_ID_ATTRIBUTE_NAME				= "id";
	public static final String HABITAT4J_PROPERTY_BEAN_DESCRIPTOR_CLASS_ATTRIBUTE_NAME			= "class";
	public static final String HABITAT4J_PROPERTY_BEAN_DESCRIPTOR_COPY_METHOD_ATTRIBUTE_NAME	= "copyMethod";

	public static final String HABITAT4J_PROPERTY_BEAN_DESCRIPTOR_COPY_METHOD_VALUE_REFLECT	= "reflect";
	public static final String HABITAT4J_PROPERTY_BEAN_DESCRIPTOR_COPY_METHOD_VALUE_SERIALIZE	= "serialize";
	public static final String HABITAT4J_PROPERTY_BEAN_DESCRIPTOR_COPY_METHOD_VALUE_DEFAULT	= HABITAT4J_PROPERTY_BEAN_DESCRIPTOR_COPY_METHOD_VALUE_REFLECT;

	public static final String HABITAT4J_PROPERTY_BEAN_ID_ATTRIBUTE_NAME					= "id";
	public static final String HABITAT4J_PROPERTY_BEAN_NAME_ATTRIBUTE_NAME					= "name";
	public static final String HABITAT4J_PROPERTY_BEAN_VALUE_ATTRIBUTE_NAME					= "value";
	public static final String HABITAT4J_PROPERTY_BEAN_INSTANCE_ATTRIBUTE_NAME				= "instance";

	public static final String HABITAT4J_PROPERTY_BEAN_INSTANCE_ATTRIBUTE_VALUE_REPLACE		= "replace";
	public static final String HABITAT4J_PROPERTY_BEAN_INSTANCE_ATTRIBUTE_VALUE_REVISE		= "revise";
	public static final String HABITAT4J_PROPERTY_BEAN_INSTANCE_ATTRIBUTE_VALUE_DEFAULT		= HABITAT4J_PROPERTY_BEAN_INSTANCE_ATTRIBUTE_VALUE_REPLACE; 

	public static final String HABITAT4J_PROPERTY_BEAN_ARRAY_INSTANCE_ATTRIBUTE_VALUE_REPLACE		= "replace";
	public static final String HABITAT4J_PROPERTY_BEAN_ARRAY_INSTANCE_ATTRIBUTE_VALUE_PREPEND		= "prepend";
	public static final String HABITAT4J_PROPERTY_BEAN_ARRAY_INSTANCE_ATTRIBUTE_VALUE_APPEND		= "append";
	public static final String HABITAT4J_PROPERTY_BEAN_ARRAY_INSTANCE_ATTRIBUTE_VALUE_DEFAULT		= HABITAT4J_PROPERTY_BEAN_ARRAY_INSTANCE_ATTRIBUTE_VALUE_REPLACE;

	public static final String HABITAT4J_PROPERTY_BEAN_HASH_INSTANCE_ATTRIBUTE_VALUE_REPLACE	= "replace";
	public static final String HABITAT4J_PROPERTY_BEAN_HASH_INSTANCE_ATTRIBUTE_VALUE_REVISE		= "revise";
	public static final String HABITAT4J_PROPERTY_BEAN_HASH_INSTANCE_ATTRIBUTE_VALUE_DEFAULT	= HABITAT4J_PROPERTY_BEAN_HASH_INSTANCE_ATTRIBUTE_VALUE_REPLACE;

	public static final String HABITAT4J_CONTEXT_APPLICATION_NAME							= "applicationName";
	
	public static final String HABITAT4J_CONTEXT_PREFIX										= "prefix";

	public static final String HABITAT4J_CONTEXT_ENVIRONMENT								= "environment";

	public static final String HABITAT4J_CONTEXT_INSTANCE_NAME								= "instanceName";
	public static final String HABITAT4J_CONTEXT_INSTANCE_DESCRIPTION						= "instanceDescription";
	public static final String HABITAT4J_CONTEXT_INSTANCE_ENUMERATION						= "instanceEnumeration";
	public static final String HABITAT4J_CONTEXT_INSTANCE_OS								= "instanceOS";

	public static final String HABITAT4J_CONTEXT_ROLE_NAME									= "roleName";
	public static final String HABITAT4J_CONTEXT_ROLE_TYPE									= "roleType";
	public static final String HABITAT4J_CONTEXT_ROLE_VENDOR								= "roleVendor";
	public static final String HABITAT4J_CONTEXT_ROLE_PRODUCT								= "roleProduct";
	public static final String HABITAT4J_CONTEXT_ROLE_VERSION								= "roleVersion";
	public static final String HABITAT4J_CONTEXT_ROLE_PATCH_LEVEL							= "rolePatchLevel";

	public static final byte HABITAT4J_PROPERTY_KIND_HABITAT4J								= 1;
	public static final byte HABITAT4J_PROPERTY_KIND_SYSTEM									= 2;
	public static final byte HABITAT4J_PROPERTY_KIND_SECURITY								= 3;	
	
	public static final String[] HABITAT4J_CONTEXT_LEGAL_ROLE_NAMES							= {
			"appServer",
			"dbServer",
			"webServer"
	};

	public static final char HABITAT4J_XML_ELEMENT_DELIMITER								= '-';
	public static final String HABITAT4J_PROPERTY_TYPE_DELIMITER							= "~";
	
	public static final String HABITAT4J_TEMP_PROPERTY_LIST_PREFIX							= "temp_";
	
	public static final String HABITAT4J_PROPERTY_ARRAY_ITEM_METHOD_PREFIX					= "setPropertyListContextPropertyArrayItem";	

	public static final String HABITAT4J_PROPERTY_HASH_ITEM_METHOD_PREFIX					= "setPropertyListContextPropertyHashItem";	
	
	public static final String HABITAT4J_PROPERTY_BEAN_INIT_METHOD_PREFIX					= "initPropertyListContextPropertyBean";

	public static final String HABITAT4J_PROPERTY_BEAN_ARRAY_INIT_METHOD_PREFIX				= "initPropertyListContextPropertyBeanArray";
	public static final String HABITAT4J_PROPERTY_BEAN_ARRAY_ITEM_INIT_METHOD_PREFIX		= "initPropertyListContextPropertyBeanArrayItem";

	public static final String HABITAT4J_PROPERTY_BEAN_HASH_INIT_METHOD_PREFIX				= "initPropertyListContextPropertyBeanHash";
	public static final String HABITAT4J_PROPERTY_BEAN_HASH_ITEM_INIT_METHOD_PREFIX			= "initPropertyListContextPropertyBeanHashItem";

	public static final String HABITAT4J_PROPERTY_BEAN_SETTER_METHOD_PREFIX					= "setPropertyListContextPropertyBean";
	
	public static final String HABITAT4J_PROPERTY_BEAN_ARRAY_SETTER_METHOD_PREFIX			= "setPropertyListContextPropertyBeanArray";
	public static final String HABITAT4J_PROPERTY_BEAN_ARRAY_ITEM_SETTER_METHOD_PREFIX		= "setPropertyListContextPropertyBeanArrayItem";

	public static final String HABITAT4J_PROPERTY_BEAN_HASH_SETTER_METHOD_PREFIX			= "setPropertyListContextPropertyBeanHash";
	public static final String HABITAT4J_PROPERTY_BEAN_HASH_ITEM_SETTER_METHOD_PREFIX		= "setPropertyListContextPropertyBeanHashItem";

	public static final String HABITAT4J_LOG4J_LOGMANAGER_CLASS								= "org.apache.log4j.LogManager";
	public static final String HABITAT4J_LOG4J_LOGMANAGER_GETLOGGER_METHOD					= "getLogger";
	public static final String HABITAT4J_LOG4J_LOGGER										= "org.productivity.java.habitat4j";
	
	public static final String EOL_SEQUENCE													= "\n";
	
	public static final String SERVER_IDENTITY_MODE_JVM_ENVIRONMENT							= "server-identity.application.environment";
	public static final String SERVER_IDENTITY_MODE_JVM_INSTANCE_NAME						= "server-identity.application.instance.name";
	public static final String SERVER_IDENTITY_MODE_JVM_INSTANCE_ENUMERATION				= "server-identity.application.instance.enumeration";
	public static final String SERVER_IDENTITY_MODE_JVM_INSTANCE_DESCRIPTION				= "server-identity.application.instance.description";
	
	public static final String SERVER_IDENTITY_MODE_FILE									= "F";
	public static final String SERVER_IDENTITY_MODE_JVM										= "J";
	public static final String SERVER_IDENTITY_MODE_NULL									= "N";
	
	public static final String SERVER_IDENTITY_INSTANCE_OS_WINDOWS							= "windows";
	public static final String SERVER_IDENTITY_INSTANCE_OS_UNIX								= "unix";
	public static final String SERVER_IDENTITY_INSTANCE_OS_OTHER							= "other";
	
	public static final String HABITAT4J_HIDDEN_FIELD_DECODED								= "##ENCODED##";
	public static final String HABITAT4J_HIDDEN_FIELD_DECRYPTED								= "##ENCRYPTED##";
	public static final String HABITAT4J_HIDDEN_FIELD_DECODED_DECRYPTED						= "##ENCODED/ENCRYPTED##";

	public static final int HABITAT4J_PROPERTY_LIST_TO_STRING_ITEMS_PER_LINE_DEFAULT		= 5;
	
	public static final String SYSTEM_PROPERTY_OS_NAME										= "os.name";
}
